File: app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.java
package ai.elimu.appstore.ui.applications;

import android.app.DownloadManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.core.content.FileProvider;
import androidx.recyclerview.widget.RecyclerView;

import java.io.File;
import java.util.List;

import ai.elimu.appstore.BaseApplication;
import ai.elimu.appstore.BuildConfig;
import ai.elimu.appstore.R;
import ai.elimu.appstore.room.entity.Application;
import ai.elimu.appstore.room.entity.ApplicationVersion;
import ai.elimu.appstore.util.FileHelper;
import ai.elimu.appstore.util.InstallationHelper;
import ai.elimu.appstore.util.SharedPreferencesHelper;
import ai.elimu.model.enums.admin.ApplicationStatus;
import timber.log.Timber;

public class ApplicationListAdapter extends RecyclerView.Adapter<ApplicationListAdapter.ApplicationViewHolder> {

    private final LayoutInflater layoutInflater;

    private final Context context;

    private List<Application> applications;

    private List<ApplicationVersion> applicationVersions;

    public ApplicationListAdapter(Context context) {
        layoutInflater = LayoutInflater.from(context);
        this.context = context;
    }

    @Override
    public ApplicationViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Timber.i("onCreateViewHolder");
        View itemView = layoutInflater.inflate(R.layout.activity_application_list_item, parent, false);
        return new ApplicationViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ApplicationViewHolder viewHolder, int position) {
        Timber.i("onBindViewHolder");
        if (applications != null) {
            // Reset button state
            viewHolder.launchButton.setVisibility(View.INVISIBLE);
            viewHolder.installButton.setVisibility(View.INVISIBLE);
            viewHolder.downloadButton.setVisibility(View.INVISIBLE);
            viewHolder.installUpdateButton.setVisibility(View.INVISIBLE);
            viewHolder.downloadUpdateButton.setVisibility(View.INVISIBLE);
            viewHolder.downloadProgressBar.setVisibility(View.INVISIBLE);

            // Populate TextViews with Application details
            Application application = applications.get(position);
            Timber.i("application.getPackageName(): \"" + application.getPackageName() + "\"");
            viewHolder.textViewFirstLine.setText(application.getPackageName());
            viewHolder.textViewSecondLine.setText(
//                    application.getApplicationStatus().toString() + ", " +
//                    application.getLiteracySkills() + ", " +
//                    application.getNumeracySkills()
                    application.getApplicationStatus().toString()
            );

            // Use 50% transparency if an Application has no corresponding APK files
            if (application.getApplicationStatus() != ApplicationStatus.ACTIVE) {
                viewHolder.textViewFirstLine.setAlpha(0.5f);
                viewHolder.textViewSecondLine.setAlpha(0.5f);
            }

            // Check if any application versions (APKs) have been uploaded to the webapp
            ApplicationVersion newestApplicationVersion = getNewestApplicationVersion(application, applicationVersions);
            Timber.i("newestApplicationVersion: " + newestApplicationVersion);
            if (newestApplicationVersion != null) {
                // Display a button matching the current state of the application
                // "Download", "Install", "Launch", "Download update", "Install update"

                if (InstallationHelper.isApplicationInstalled(application.getPackageName(), context)) {
                    // The APK has been installed

                    // Check if an update is available for download
                    int versionCodeInstalled = InstallationHelper.getVersionCodeOfInstalledApplication(application.getPackageName(), context);
                    Timber.i("versionCodeInstalled: " + versionCodeInstalled);
                    ApplicationVersion applicationVersion = getNewestApplicationVersion(application, applicationVersions);
                    Timber.i("applicationVersion.getVersionCode(): " + applicationVersion.getVersionCode());
                    if (versionCodeInstalled < applicationVersion.getVersionCode()) {
                        // An update is available for download

                        // If the APK has been downloaded (but not yet installed), display the "Install update" button
                        File apkFile = FileHelper.getApkFile(application.getPackageName(), applicationVersion.getVersionCode(), context);
                        Timber.i("apkFile: " + apkFile);
                        Timber.i("apkFile.exists(): " + apkFile.exists());
                        if (apkFile.exists()) {
                            viewHolder.installUpdateButton.setVisibility(View.VISIBLE);
                            View.OnClickListener onClickListener = v -> {
                                Timber.i("viewHolder.installUpdateButton onClick");

                                // Initiate installation of the APK file
                                IntentFilter intentFilter = new IntentFilter(Intent.ACTION_PACKAGE_ADDED);
                                intentFilter.addDataScheme("package");
                                context.registerReceiver(new PackageAddedReceiver(position), intentFilter);
                                Uri apkUri = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".apk.provider", apkFile);
                                Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);
                                intent.setData(apkUri);
                                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                                context.startActivity(intent);
                            };
                            viewHolder.installUpdateButton.setOnClickListener(onClickListener);
                        }

                        // If the APK has not been downloaded, display the "Download update" button
                        if (!apkFile.exists()) {
                            viewHolder.downloadUpdateButton.setVisibility(View.VISIBLE);
                            ApplicationVersion finalApplicationVersion = applicationVersion;
                            viewHolder.downloadUpdateButton.setOnClickListener(v -> {
                                Timber.i("viewHolder.downloadUpdateButton onClick");

                                // Initiate download of the APK file
                                context.registerReceiver(new DownloadCompleteReceiver(position), new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
                                BaseApplication baseApplication = (BaseApplication) context.getApplicationContext();
                                String fileUrl = baseApplication.getBaseUrl() + finalApplicationVersion.getFileUrl();
                                Timber.i("fileUrl: " +  fileUrl);
                                DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
                                DownloadManager.Request request = new DownloadManager.Request(Uri.parse(fileUrl));
                                String destinationInExternalFilesDir = File.separator + "lang-" + SharedPreferencesHelper.getLanguage(context).getIsoCode() + File.separator + "apks" + File.separator + apkFile.getName();
                                Timber.i("destinationInExternalFilesDir: " +  destinationInExternalFilesDir);
                                request.setDestinationInExternalFilesDir(context, null, destinationInExternalFilesDir);
                                long downloadId = downloadManager.enqueue(request);
                                Timber.i("downloadId: " +  downloadId);

                                // Replace download button with progress bar
                                viewHolder.downloadUpdateButton.setVisibility(View.INVISIBLE);
                                viewHolder.downloadProgressBar.setVisibility(View.VISIBLE);
                            });
                        }
                    } else {
                        // The installed APK is up-to-date

                        // Display the "Launch" button
                        viewHolder.launchButton.setVisibility(View.VISIBLE);
                        viewHolder.launchButton.setOnClickListener((View.OnClickListener) v -> {
                            Timber.i("onClick");

                            Timber.i("Launching \"" + application.getPackageName() + "\"");
                            PackageManager packageManager = context.getPackageManager();
                            Intent launchIntent = packageManager.getLaunchIntentForPackage(application.getPackageName());
                            Timber.i("launchIntent: " + launchIntent);
                            context.startActivity(launchIntent);
                        });
                    }
                } else {
                    // The APK has not been installed

                    // Fetch information about the newest APK file
                    ApplicationVersion applicationVersion = getNewestApplicationVersion(application, applicationVersions);
                    if (applicationVersion == null) {
                        return;
                    }
                    Timber.i("applicationVersion.getVersionCode(): " + applicationVersion.getVersionCode());

                    // If the APK has been downloaded (but not yet installed), display the "Install" button
                    File apkFile = FileHelper.getApkFile(application.getPackageName(), applicationVersion.getVersionCode(), context);
                    Timber.i("apkFile: " + apkFile);
                    Timber.i("apkFile.exists(): " + apkFile.exists());
                    if (apkFile.exists()) {
                        viewHolder.installButton.setVisibility(View.VISIBLE);
                        View.OnClickListener onClickListener = v -> {
                            Timber.i("viewHolder.installButton onClick");

                            // Initiate installation of the APK file
                            IntentFilter intentFilter = new IntentFilter(Intent.ACTION_PACKAGE_ADDED);
                            intentFilter.addDataScheme("package");
                            context.registerReceiver(new PackageAddedReceiver(position), intentFilter);
                            Uri apkUri = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".apk.provider", apkFile);
                            Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);
                            intent.setData(apkUri);
                            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                            context.startActivity(intent);
                        };
                        viewHolder.installButton.setOnClickListener(onClickListener);
                    }

                    // If the APK has not been downloaded, display the "Download" button
                    if (!apkFile.exists()) {
                        viewHolder.downloadButton.setVisibility(View.VISIBLE);
                        ApplicationVersion finalApplicationVersion = applicationVersion;
                        viewHolder.downloadButton.setOnClickListener(v -> {
                            Timber.i("viewHolder.downloadButton onClick");

                            // Initiate download of the APK file
                            context.registerReceiver(new DownloadCompleteReceiver(position), new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
                            BaseApplication baseApplication = (BaseApplication) context.getApplicationContext();
                            String fileUrl = baseApplication.getBaseUrl() + finalApplicationVersion.getFileUrl();
                            Timber.i("fileUrl: " +  fileUrl);
                            DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
                            DownloadManager.Request request = new DownloadManager.Request(Uri.parse(fileUrl));
                            String destinationInExternalFilesDir = File.separator + "lang-" + SharedPreferencesHelper.getLanguage(context).getIsoCode() + File.separator + "apks" + File.separator + apkFile.getName();
                            Timber.i("destinationInExternalFilesDir: " +  destinationInExternalFilesDir);
                            request.setDestinationInExternalFilesDir(context, null, destinationInExternalFilesDir);
                            long downloadId = downloadManager.enqueue(request);
                            Timber.i("downloadId: " +  downloadId);

                            // Replace download button with progress bar
                            viewHolder.downloadButton.setVisibility(View.INVISIBLE);
                            viewHolder.downloadProgressBar.setVisibility(View.VISIBLE);
                        });
                    }
                }
            }
        }
    }

    private ApplicationVersion getNewestApplicationVersion(Application application, List<ApplicationVersion> applicationVersions) {
        ApplicationVersion applicationVersion = null;
        for (ApplicationVersion appVersion : applicationVersions) {
            if (appVersion.getApplicationId() == application.getId()) {
                applicationVersion = appVersion;
                break;
            }
        }
        return applicationVersion;
    }

    @Override
    public int getItemCount() {
        Timber.i("getItemCount");
        if (applications == null) {
            return 0;
        } else {
            return applications.size();
        }
    }

    public void setApplications(List<Application> applications) {
        this.applications = applications;
    }

    public void setApplicationVersions(List<ApplicationVersion> applicationVersions) {
        this.applicationVersions = applicationVersions;
    }


    class ApplicationViewHolder extends RecyclerView.ViewHolder {

        private final TextView textViewFirstLine;
        private final TextView textViewSecondLine;

        private Button launchButton;
        private Button installButton;
        private Button downloadButton;
        private Button installUpdateButton;
        private Button downloadUpdateButton;
        private ProgressBar downloadProgressBar;

        private ApplicationViewHolder(View itemView) {
            super(itemView);
            Timber.i("ApplicationViewHolder");

            textViewFirstLine = itemView.findViewById(R.id.textViewFirstLine);
            textViewSecondLine = itemView.findViewById(R.id.textViewSecondLine);

            launchButton = itemView.findViewById(R.id.list_item_launch_button);
            installButton = itemView.findViewById(R.id.list_item_install_button);
            downloadButton = itemView.findViewById(R.id.list_item_download_button);
            installUpdateButton = itemView.findViewById(R.id.list_item_install_update_button);
            downloadUpdateButton = itemView.findViewById(R.id.list_item_download_update_button);
            downloadProgressBar = itemView.findViewById(R.id.list_item_download_progressbar);
        }
    }


    private class DownloadCompleteReceiver extends BroadcastReceiver {

        private int itemPosition;

        public DownloadCompleteReceiver(int itemPosition) {
            this.itemPosition = itemPosition;
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            Timber.i("onReceive");

            Timber.i("intent: " + intent);
            notifyItemChanged(itemPosition);
        }
    }


    private class PackageAddedReceiver extends BroadcastReceiver {

        private int itemPosition;

        public PackageAddedReceiver(int itemPosition) {
            this.itemPosition = itemPosition;
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            Timber.i("onReceive");

            Timber.i("intent: " + intent);
            Timber.i("intent.getData(): " + intent.getData());
            notifyItemChanged(itemPosition);
        }
    }
}

File: app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListAdapter.kt
package ai.elimu.appstore.ui.applications

import ai.elimu.appstore.BuildConfig
import ai.elimu.appstore.R
import ai.elimu.appstore.room.entity.Application
import ai.elimu.appstore.room.entity.ApplicationVersion
import ai.elimu.appstore.ui.applications.ApplicationListAdapter.ApplicationViewHolder
import ai.elimu.appstore.util.FileHelper
import ai.elimu.appstore.util.FileHelper.getApkFile
import ai.elimu.appstore.util.InstallationHelper.getVersionCodeOfInstalledApplication
import ai.elimu.appstore.util.InstallationHelper.isApplicationInstalled
import ai.elimu.appstore.util.SharedPreferencesHelper.getLanguage
import ai.elimu.model.v2.enums.admin.ApplicationStatus
import android.app.DownloadManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.recyclerview.widget.RecyclerView
import timber.log.Timber
import java.io.File
import androidx.core.net.toUri
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ApplicationListAdapter(
    private val context: Context,
    private val downloadReceiver: DownloadCompleteReceiver
) :
    RecyclerView.Adapter<ApplicationViewHolder>() {
    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)

    private var applications: List<Application>? = null

    private var applicationVersions: List<ApplicationVersion>? = null
    private val registeredReceivers: MutableSet<BroadcastReceiver> by lazy { mutableSetOf() }

    private val downloadManager =
        context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ApplicationViewHolder {
        Timber.i("onCreateViewHolder")
        val itemView =
            layoutInflater.inflate(R.layout.activity_application_list_item, parent, false)
        return ApplicationViewHolder(itemView)
    }

    override fun onBindViewHolder(viewHolder: ApplicationViewHolder, position: Int) {
        Timber.i("onBindViewHolder")
        if (applications != null) {
            // Reset button state
            viewHolder.launchButton.visibility = View.INVISIBLE
            viewHolder.installButton.visibility = View.INVISIBLE
            viewHolder.downloadButton.visibility = View.INVISIBLE
            viewHolder.installUpdateButton.visibility = View.INVISIBLE
            viewHolder.downloadUpdateButton.visibility = View.INVISIBLE
            viewHolder.downloadProgressBar.visibility = View.INVISIBLE

            // Populate TextViews with Application details
            val application = applications!![position]
            Timber.i("application.getPackageName(): ${application.packageName}")
            viewHolder.textViewFirstLine.text = application.packageName
            viewHolder.textViewSecondLine.text = application.applicationStatus.toString()

            // Use 50% transparency if an Application has no corresponding APK files
            if (application.applicationStatus != ApplicationStatus.ACTIVE) {
                viewHolder.textViewFirstLine.alpha = 0.5f
                viewHolder.textViewSecondLine.alpha = 0.5f
            }

            // Check if any application versions (APKs) have been uploaded to the webapp
            val newestApplicationVersion = getNewestApplicationVersion(
                application,
                applicationVersions!!
            )
            Timber.i("newestApplicationVersion: $newestApplicationVersion")
            if (newestApplicationVersion != null) {
                // Display a button matching the current state of the application
                // "Download", "Install", "Launch", "Download update", "Install update"

                if (isApplicationInstalled(application.packageName, context)) {
                    // The APK has been installed

                    // Check if an update is available for download

                    val versionCodeInstalled =
                        getVersionCodeOfInstalledApplication(application.packageName, context)
                    Timber.i("versionCodeInstalled: $versionCodeInstalled")
                    val applicationVersion = getNewestApplicationVersion(
                        application,
                        applicationVersions!!
                    )
                    Timber.i(
                        "applicationVersion.getVersionCode(): %s",
                        applicationVersion!!.versionCode
                    )
                    if (versionCodeInstalled < applicationVersion.versionCode) {
                        // An update is available for download

                        // If the APK has been downloaded (but not yet installed), display the "Install update" button

                        val apkFile = getApkFile(
                            application.packageName,
                            applicationVersion.versionCode,
                            context
                        )
                        Timber.i("apkFile: $apkFile")
                        Timber.i("apkFile.exists(): %s", apkFile!!.exists())
                        if (apkFile.exists()) {
                            viewHolder.installUpdateButton.visibility = View.VISIBLE
                            val onClickListener = View.OnClickListener { _: View? ->
                                Timber.i("viewHolder.installUpdateButton onClick")
                                // Initiate installation of the APK file
                                val intentFilter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
                                intentFilter.addDataScheme("package")
                                val packageAddedReceiver = PackageAddedReceiver(position)
                                context.registerReceiver(
                                    packageAddedReceiver,
                                    intentFilter
                                )
                                registeredReceivers.add(packageAddedReceiver)
                                val apkUri = FileProvider.getUriForFile(
                                    context, BuildConfig.APPLICATION_ID + ".apk.provider",
                                    apkFile
                                )
                                val intent = Intent(Intent.ACTION_INSTALL_PACKAGE)
                                intent.setData(apkUri)
                                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                                context.startActivity(intent)
                            }
                            viewHolder.installUpdateButton.setOnClickListener(onClickListener)
                        }

                        // If the APK has not been downloaded, display the "Download update" button
                        if (!apkFile.exists()) {
                            viewHolder.downloadUpdateButton.visibility = View.VISIBLE
                            viewHolder.downloadUpdateButton.setOnClickListener {
                                Timber.i("viewHolder.downloadUpdateButton onClick")
                                // Initiate download of the APK file
                                Timber.i("finalApplicationVersion.fileUrl: ${applicationVersion.fileUrl}")
                                val request =
                                    DownloadManager.Request(applicationVersion.fileUrl.toUri())
                                val destinationInExternalFilesDir =
                                    File.separator + "lang-" + getLanguage(
                                        context
                                    )!!
                                        .isoCode + File.separator + "apks" + File.separator + apkFile.name
                                request.setDestinationInExternalFilesDir(
                                    context,
                                    null,
                                    destinationInExternalFilesDir
                                )
                                val downloadId = downloadManager.enqueue(request)
                                if (downloadId != -1L) {
                                    downloadReceiver.addDownloadListener(downloadId) {
                                        handleDownloadComplete(itemPosition = position,
                                            relativeFilePath = destinationInExternalFilesDir,
                                            checkSum = applicationVersion.checksumMd5
                                        )
                                    }
                                    Timber.i("downloadId: $downloadId")
                                }

                                // Replace download button with progress bar
                                viewHolder.downloadUpdateButton.visibility =
                                    View.INVISIBLE
                                viewHolder.downloadProgressBar.visibility =
                                    View.VISIBLE
                            }
                        }
                    } else {
                        // The installed APK is up-to-date

                        // Display the "Launch" button

                        viewHolder.launchButton.visibility = View.VISIBLE
                        viewHolder.launchButton.setOnClickListener {
                            Timber.i("onClick")
                            Timber.i("Launching ${application.packageName}")
                            val packageManager = context.packageManager
                            val launchIntent =
                                packageManager.getLaunchIntentForPackage(application.packageName)
                            Timber.i("launchIntent: $launchIntent")
                            context.startActivity(launchIntent)
                        }
                    }
                } else {
                    // The APK has not been installed

                    // Fetch information about the newest APK file

                    val applicationVersion = getNewestApplicationVersion(
                        application,
                        applicationVersions!!
                    )
                    if (applicationVersion == null) {
                        return
                    }
                    Timber.i(
                        "applicationVersion.getVersionCode(): %s",
                        applicationVersion.versionCode
                    )

                    // If the APK has been downloaded (but not yet installed), display the "Install" button
                    val apkFile =
                        getApkFile(application.packageName, applicationVersion.versionCode, context)
                    Timber.i("apkFile: $apkFile")
                    Timber.i("apkFile.exists(): %s", apkFile!!.exists())
                    if (apkFile.exists()) {
                        viewHolder.installButton.visibility = View.VISIBLE
                        val onClickListener = View.OnClickListener { _: View? ->
                            Timber.i("viewHolder.installButton onClick")
                            // Initiate installation of the APK file
                            val intentFilter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
                            intentFilter.addDataScheme("package")
                            val packageAddedReceiver = PackageAddedReceiver(position)
                            context.registerReceiver(packageAddedReceiver, intentFilter)
                            val apkUri = FileProvider.getUriForFile(
                                context, BuildConfig.APPLICATION_ID + ".apk.provider",
                                apkFile
                            )
                            registeredReceivers.add(packageAddedReceiver)
                            val intent = Intent(Intent.ACTION_INSTALL_PACKAGE)
                            intent.setData(apkUri)
                            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                            context.startActivity(intent)
                        }
                        viewHolder.installButton.setOnClickListener(onClickListener)
                    }

                    // If the APK has not been downloaded, display the "Download" button
                    if (!apkFile.exists()) {
                        viewHolder.downloadButton.visibility = View.VISIBLE
                        val finalApplicationVersion: ApplicationVersion = applicationVersion
                        viewHolder.downloadButton.setOnClickListener {
                            Timber.i("viewHolder.downloadButton onClick")
                            // Initiate download of the APK file
                            Timber.i("finalApplicationVersion.fileUrl: ${finalApplicationVersion.fileUrl}")
                            val request =
                                DownloadManager.Request(finalApplicationVersion.fileUrl.toUri())
                            val destinationInExternalFilesDir =
                                File.separator + "lang-" + getLanguage(
                                    context
                                )!!
                                    .isoCode + File.separator + "apks" + File.separator + apkFile.name
                            Timber.i("destinationInExternalFilesDir: $destinationInExternalFilesDir")

                            request.setDestinationInExternalFilesDir(
                                context,
                                null,
                                destinationInExternalFilesDir
                            )
                            val downloadId = downloadManager.enqueue(request)
                            downloadReceiver.addDownloadListener(downloadId) {
                                handleDownloadComplete(itemPosition = position,
                                    relativeFilePath = destinationInExternalFilesDir,
                                    checkSum = applicationVersion.checksumMd5
                                )
                            }

                            Timber.i("downloadId: $downloadId")

                            // Replace download button with progress bar
                            viewHolder.downloadButton.visibility = View.INVISIBLE
                            viewHolder.downloadProgressBar.visibility = View.VISIBLE
                        }
                    }
                }
            }
        }
    }

    private fun getNewestApplicationVersion(
        application: Application,
        applicationVersions: List<ApplicationVersion>,
    ): ApplicationVersion? {
        var applicationVersion: ApplicationVersion? = null
        for (appVersion in applicationVersions) {
            if (appVersion.applicationId == application.id) {
                applicationVersion = appVersion
                break
            }
        }
        return applicationVersion
    }

    private fun handleDownloadComplete(
        itemPosition: Int,
        relativeFilePath: String,
        checkSum: String
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val downloadedFile = context.getExternalFilesDir(relativeFilePath)
            val downloadedFileCheckSum =
                FileHelper.calculateMD5Checksum(downloadedFile?.absolutePath ?: "")
            withContext(Dispatchers.Main) {
                if (downloadedFileCheckSum != checkSum) {
                    downloadedFile?.delete()
                    val activity = context as? ApplicationListActivity ?: return@withContext
                    Snackbar.make(
                        activity.window.decorView,
                        context.getString(R.string.download_error),
                        Snackbar.LENGTH_LONG
                    ).setBackgroundTint(ContextCompat.getColor(context, R.color.red)).show()
                }
                notifyItemChanged(itemPosition)
            }
        }
    }

    override fun getItemCount(): Int {
        Timber.i("getItemCount")
        return if (applications == null) {
            0
        } else {
            applications!!.size
        }
    }

    fun setApplications(applications: List<Application>?) {
        this.applications = applications
    }

    fun setApplicationVersions(applicationVersions: List<ApplicationVersion>?) {
        this.applicationVersions = applicationVersions
    }

    fun unregisterReceiver(context: Context) {
        for (receiver in registeredReceivers) {
            try {
                context.unregisterReceiver(receiver)
            } catch (e: IllegalArgumentException) {
                Timber.e(e)
            }
        }
    }

    inner class ApplicationViewHolder internal constructor(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        val textViewFirstLine: TextView
        val textViewSecondLine: TextView

        val launchButton: Button
        val installButton: Button
        val downloadButton: Button
        val installUpdateButton: Button
        val downloadUpdateButton: Button
        val downloadProgressBar: ProgressBar

        init {
            Timber.i("ApplicationViewHolder")

            textViewFirstLine = itemView.findViewById(R.id.textViewFirstLine)
            textViewSecondLine = itemView.findViewById(R.id.textViewSecondLine)

            launchButton = itemView.findViewById(R.id.list_item_launch_button)
            installButton = itemView.findViewById(R.id.list_item_install_button)
            downloadButton = itemView.findViewById(R.id.list_item_download_button)
            installUpdateButton = itemView.findViewById(R.id.list_item_install_update_button)
            downloadUpdateButton = itemView.findViewById(R.id.list_item_download_update_button)
            downloadProgressBar = itemView.findViewById(R.id.list_item_download_progressbar)
        }
    }

    private inner class PackageAddedReceiver(private val itemPosition: Int) : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            Timber.i("onReceive")

            Timber.i("intent: $intent")
            Timber.i("intent.getData(): %s", intent.data)
            notifyItemChanged(itemPosition)
            context.unregisterReceiver(this)
            registeredReceivers.remove(this)
        }
    }
}
